0.0.1/agent.py

```diff
@@ -1,63 +1,99 @@
         for message in reversed(chat_history):
-            match = USD_ALLOWANCE_GOAL_REGEX.match(message['content'])
-            if message['role'] == 'user' and match:
+            match = USD_ALLOWANCE_GOAL_REGEX.match(message["content"])
+            if message["role"] == "user" and match:
                 return match.group(1)
-        return ''
+        return ""
 
     def get_last_search_term(self, chat_history):
         for message in reversed(chat_history):
-           if message['role'] == 'user':
-               return message['content']
-        return ''
+            if message["role"] == "user":
+                return message["content"]
+        return ""
 
     def run(self):
         # A system message guides an agent to solve specific tasks.
-        prompt = {"role": "system", "content": "You are an assistant that helps people set goals for growth in the USD value of their crypto assets such that when that percentage in growth has been reached or surpassed, you look at their tokens and determine the tokens and quantities of each to swap for USDT stablecoins or USDC stablecoins"}
+        prompt = {
+            "role": "system",
+            "content": """You are an assistant that helps people set goals for growth in the USD value of their crypto assets
+            such that when that percentage in growth has been reached or surpassed, you look at their tokens and
+            determine the tokens and quantities of each to swap for USDT stablecoins or USDC stablecoins.
+            You also have access to some auxiliary info about their wallets and accounts""",
+        }
 
         # Use the model set in the metadata to generate a response
-        result = self.env.completion([prompt] + self.env.list_messages())
+        # result = self.env.completion([prompt] + self.env.list_messages())
+        tools = self.env.get_tool_registry().get_all_tool_definitions()
+        res = self.env.completion_and_get_tools_calls(
+            [prompt] + [self.env.get_last_message() or ""], tools=tools
+        )
+        self.env.add_system_log(f"Should call tools: {res.tool_calls}", logging.DEBUG)
+
+        if res.message:
+            self.env.add_reply(res.message)
+        if res.tool_calls and len(res.tool_calls) > 0:
+            tool_call_results = self._handle_tool_calls(res.tool_calls)
+            if len(tool_call_results) > 0:
+                self.env.add_system_log(
+                    f"Got tool call results: {tool_call_results}", logging.DEBUG
+                )
+                result = self.env.completion(
+                    [prompt] + self.env.list_messages() + tool_call_results
+                )
 
         chat_history = self.env.list_messages()
         last_user_query = self.get_last_search_term(chat_history)
-        if last_user_query == "recommend swaps" or last_user_query == "suggest tokens to swap":
-          result = self.recommend_token_allocations_to_swap_for_stablecoins()
+        if (
```

Comment:
These old ones continue to work but if we agree on this direction I can refactor these to be function driven too

Reply:
I'm onboard with the extensible function call approach 👍🏾